configs/provider_validation.go

```diff
@@ -0,0 +1,246 @@
+
+		if !(localName || configAlias || emptyConfig) {
+			severity := hcl.DiagError
+
+			// we still allow default configs, so switch to a warning if the incoming provider is a default
+			if providerAddr.Provider.IsDefault() {
+				severity = hcl.DiagWarning
+			}
+
+			diags = append(diags, &hcl.Diagnostic{
+				Severity: severity,
+				Summary:  fmt.Sprintf("Provider %s is undefined", name),
+				Detail: fmt.Sprintf("Module %s does not declare a provider named %s.\n", cfg.Path, name) +
+					fmt.Sprintf("If you wish to specify a provider configuration for the module, add an entry for %s in the required_providers block within the module.", name),
+				Subject: &passed.InChild.NameRange,
+			})
+		}
+
+		// The provider being passed in must also be of the correct type.
+		// While we would like to ensure required_providers exists here,
+		// implied default configuration is still allowed.
+		pTy := addrs.NewDefaultProvider(passed.InParent.Name)
+
+		// use the full address for a nice diagnostic output
+		parentAddr := addrs.AbsProviderConfig{
+			Module:   cfg.Parent.Path,
+			Provider: pTy,
+			Alias:    passed.InParent.Alias,
+		}
+
+		if cfg.Parent.Module.ProviderRequirements != nil {
+			req, defined := cfg.Parent.Module.ProviderRequirements.RequiredProviders[name]
+			if defined {
+				parentAddr.Provider = req.Type
+			}
+		}
+
+		if !providerAddr.Provider.Equals(parentAddr.Provider) {
+			diags = append(diags, &hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  fmt.Sprintf("Invalid type for provider %s", providerAddr),
+				Detail: fmt.Sprintf("Cannot use configuration from %s for %s.\n", parentAddr, providerAddr) +
+					"The given provider configuration is for a different provider type.",
+				Subject: &passed.InChild.NameRange,
+			})
+		}
+	}
+
+	// Empty configurations are no longer needed
+	for name, src := range emptyConfigs {
+		detail := fmt.Sprintf("Remove the %s provider block from %s.", name, cfg.Path)
+
+		isAlias := strings.Contains(name, ".")
+		_, isConfigAlias := configAliases[name]
+		_, isLocalName := localNames[name]
+
+		if isAlias && !isConfigAlias {
+			localName := strings.Split(name, ".")[0]
+			detail = fmt.Sprintf("Remove the %s provider block from %s. Add %s to the list of configuration_aliases for %s in required_providers to define the provider configuration name.", name, cfg.Path, name, localName)
+		}
+
+		if !isAlias && !isLocalName {
+			// if there is no local name, add a note to include it in the
+			// required_provider block
+			detail += fmt.Sprintf("\nTo ensure the correct provider configuration is used, add %s to the required_providers configuration", name)
+		}
+
+		diags = append(diags, &hcl.Diagnostic{
+			Severity: hcl.DiagWarning,
+			Summary:  "Empty provider configuration blocks are not required",
+			Detail:   detail,
+			Subject:  src,
+		})
+	}
+
+	if diags.HasErrors() {
+		return diags
+	}
+
+	// If we have gotten this far without errors, verify that any passed in
```

Comment:
Either you're missing a verification step or forgot to remove a dangling comment :)